### ISSUE TITLE
[New Hub] PaleoCube and PaleoHack hubs

### ISSUE THREAD

**Issue by colliand** [2i2c-org/infrastructure#1418]
### Hub Description

2i2c has been engaged by members of the paleoclimate community. Deborah Khider (@khider) has been our principal contact in establishing this connection. This engagement extends previous work we've done to support the PaleoHackWeeks. We will learn more from Deborah about the community and their needs as we roll out the service. Here are a few links that provide some background information. 

[LinkedEarth](http://linked.earth/)

[Deborah's home page](https://earth.usc.edu/~khider/)

### Community Representative(s)

@khider

### Important dates

- **Target start date**: June 30, 2022
- **Required start date**: July 15, 2022
- **Any important dates for usage**: _???_

### Hub Authentication Type

CILogon (e.g., username@institution.org)

### Hub logo information

- **URL to Hub Image**: https://github.com/LinkedEarth/Logos/blob/master/linkedearth_hub_logo.png?rgh-link-date=2022-06-15T16%3A21%3A32Z
- **URL for Image Link**: https://linked.earth/

### Hub user image

- Repository for user image: https://quay.io/repository/linkedearth/pyleoclim
- User image registry: quay.io
- User image tag and name: `latest` tags

Once again, seeking input from @khider! 

Does your community have a Docker Image that specifies the software environment you wish to use in your hub(s)? If so, please share a link and we will review and take steps to set things up using that environment. If not, we can proceed with a Pangeo image and help you and your community adapt the environment to better meet your needs. 

### Extra features you'd like to enable

- [x] Specific cloud provider or datacenter: GCP, us-central1-c
- [x] Dedicated Kubernetes cluster
- [x] Scalable Dask Cluster

### Hub URL

linkedearth.2i2c.cloud

### Hub Type

daskhub

### Tasks to deploy the hub

- [x] Engineer who will deploy the hub is assigned
- [x] Deploy information filled in above
- [x] Initial cluster deployment PR: https://github.com/2i2c-org/infrastructure/pull/1476
- [x] Initial Hub deployment PR: https://github.com/2i2c-org/infrastructure/pull/1476
- [x] Administrators able to log on
- [x] Community Representative satisfied with hub environment
- [x] Hub now in steady-state

**Comment by damianavila** [comment]
We already have a live PaleoHack hub: https://github.com/2i2c-org/infrastructure/blob/master/config/clusters/2i2c/paleohack2021.values.yaml

Is this a request for a new one, @colliand?

**Comment by colliand** [comment]
Yes, I believe this request is for a persistent hub. I look forward to further guidance from @khider.

**Comment by khider** [comment]
Yes, this is a request for a new, permanent one for research purposes. Answering/Commenting on above:

1. Authentification: any chance to also allow for Github auth?

2. Hub links and logos:
link: linkedearth.2i2c.cloud
Image: https://github.com/LinkedEarth/Logos/blob/master/linkedearth_hub_logo.png

3. Image and repo. The idea is to allow to pull examples from several repositories (which may not have the same environment). 
Repos gallery (click image link to get to the GitHub repos): http://linked.earth/gallery.html#gallery
There are images available here: https://quay.io/repository/linkedearth/pyleoclim?tab=tags

4. We're working with Pangeo Forge to transform some simulations from netcdf to Zarr format. Task scaling may be needed once this task is completed and some of our example notebooks would use the results of these simulations. 

5. Cloud credit: we may need to have a discussion. For the grant, I had to separate the 2i2c service from cloud billing. So I need to setup an account. Guidance would be helpful for this.

**Comment by sgibson91** [comment]
> any chance to also allow for Github auth?

I don't believe it's possible to mix authentication types presently, so it would have to be one or the other.

**Comment by damianavila** [comment]
> Cloud credit: we may need to have a discussion. For the grant, I had to separate the 2i2c service from cloud billing. So I need to setup an account. Guidance would be helpful for this.

Is there any cloud provider preference?

**Comment by khider** [comment]
> > any chance to also allow for Github auth?
> 
> I don't believe it's possible to mix authentication types presently, so it would have to be one or the other.

Can they use their own institution logon or would have to somehow register with us?

**Comment by khider** [comment]
> > Cloud credit: we may need to have a discussion. For the grant, I had to separate the 2i2c service from cloud billing. So I need to setup an account. Guidance would be helpful for this.
> 
> Is there any cloud provider preference?

No. So open to discussion on what would be best.

Just one thing to keep in mind: some of the data will be on Pangeo-forge. So we either need to get close to it or duplicate for the life of this project.

**Comment by colliand** [comment]
I spoke with @khider via telephone. Here is a brief report and a few requests for action from the 2i2c team:

1. USC rules require the cloud billing account for the `livingearth.2i2c.cloud` hub to be held by USC and not by 2i2c. The reasons for this involve accounting and billing overhead. Deborah is prepared to swipe a credit card to establish the cloud billing account and share credentials with the 2i2c. 
2. The livingearth community will use CMIP data (Petabytes) hosted on Pangeo Forge. Deborah requests that the compute be located near the data. With which cloud vendor and in which data center should 2i2c set up this hub? I call out @rabernat for guidance and @damianavila for following up with @khider to secure access to the cloud infrastructure.
3. This community uses a collection of software environments. There were significant changes when their software changed from version 0.7 to 0.8. As a result, some useful notebooks will not run when the base is at the 0.8 level. Deborah asked if 2i2c can make it easy for users to swap out software environments. I understand that @fperez and @rabernat are exploring similar ideas so thought it wise to share this request. 

The main next actions:

- [ ] Determine which vendor and data center should be used to host the `livingearth.2i2c.cloud` hub.
- [ ] Establish the cloud billing account
- [ ] Share credentials to allow the 2i2c team to access the cloud account
- [ ] Stand up the hub and do awesome science!

**Comment by damianavila** [comment]
@rabernat do you have any recommendations about where to deploy this cluster/hub given the fact they are trying to use data hosted in Pangeo Forge? cc @yuvipanda who might have thoughts as well.

**Comment by damianavila** [comment]
Assuming this is the data we are talking about: https://pangeo-forge.org/dashboard/feedstock/7, @khider can you confirm if this is the data we are talking about?
It seems it is handled by this bakery: pangeo-ldeo-nsf-earthcube and this one seems to be a GCS one: https://github.com/pangeo-forge/pangeo-forge-gcs-bakery.
So, it might be the case the data actually lives in GCS US central-1?
I am not familiarized enough with Pangeo Forge... so it might be the case I am fundamentally wrong 😉. 

cc @sgibson91 (it might be the case this is actually a GCP hub instead of an AWS, but not sure yet).

**Comment by khider** [comment]
Yes, for the modern data. We also need to work with them to get the PMIP data onto Pangeo Forge, I assume at the same location.

**Comment by rabernat** [comment]
The data are actually being deposited in the OSN storage pod at NCSA, so not actual in any commercial cloud. They are fully public, with no requestor pays or other authorization required. I would place the hub in whichever cloud region is in closest network proximity to NCSA. Google cloud US-CENTRAL-1 has worked well for us in the past, but I'm sure there is a nearby AWS region too.

**Comment by damianavila** [comment]
> Google cloud US-CENTRAL-1 has worked well for us in the past

OK, given this information, a GCP-based daskhub deployment in their own dedicated cluster seems to be the right choice.

@yuvipanda @sgibson91, do you agree with me on this assessment (since you have been lately interacting with new dedicated clusters on GCP, I think you may have good feedback here)?

**Comment by yuvipanda** [comment]
Yep, putting this on gcp at us-central-1 is the way to go. Let's put it on us-central1-c as that has most variety of GPU available (https://cloud.google.com/compute/docs/gpus/gpu-regions-zones) in case we need it in the future.

**Comment by damianavila** [comment]
OK, @khider, I think the next step would be to follow: https://docs.2i2c.org/en/latest/admin/howto/create-billing-account.html#billing-and-cloud-accounts, to create the GCP cloud billing account and add 2i2c engs to it.

**Comment by damianavila** [comment]
Btw, in step 2.5, could you please add the following emails: yuvipanda@2i2c.org, sgibson@2i2c.org, georgianaelena@2i2c.org, and damianavila@2i2c.org. Thanks!

**Comment by khider** [comment]
And done! Everyone should now have access.

**Comment by damianavila** [comment]
For contextual/additional information about the above assignations, @sgibson91 will be the lead developer for this hub with @yuvipanda's assistance as a secondary companion/helper/supporter.

**Comment by sgibson91** [comment]
@khider to confirm, is the billing account you created called "LinkedEarth Jupyter Hub"?

**Comment by sgibson91** [comment]
I have created the `linked-earth-hubs` project in our 2i2c.org GCP organgisation, attached the Linked Earth Jupyter Hub billing account, and given everyone on the engineering team access to the new project

**Comment by sgibson91** [comment]
@khider can you confirm if you would prefer CILogon authentication (username@institution.edu) or GitHub authentication please?

**Comment by khider** [comment]
> @khider can you confirm if you would prefer CILogon authentication (`username@institution.edu`) or GitHub authentication please?

CILogon preferred

**Comment by sgibson91** [comment]
Thanks! And what's the institution domain please? The institution.something part?

**Comment by sgibson91** [comment]
The hubs are located at:

- https://staging.linkedearth.2i2c.cloud
- https://linkedearth.2i2c.cloud

**Comment by khider** [comment]
> We have also setup a daskhub for you, but dask_gateway is not installed in the pyleoclim image you asked us to use, so you may have trouble using dask in a scalable fashion until dask_gateway is available in the image
> 
> <img alt="Screenshot 2022-06-30 at 13 53 38" width="672" src="https://user-images.githubusercontent.com/44771837/176681755-627a13f8-35f1-4077-be10-ca3f1aa0722e.png">

I can add it to the image, and since we're using the latest tag, it should pull automatically, correct? Which version should I use? It seems like I need to same as what's on the server according to: https://gateway.dask.org/install-user.html

**Comment by sgibson91** [comment]
@khider, I need to know your `institution.domain`  (e.g. `berkeley.edu`) so I can add it to the CILogon config. At the minute, only folk with 2i2c.org Google accounts can log in.

I will find out what version of dask_gateway you need to install into your image

**Comment by sgibson91** [comment]
> we're using the latest tag, it should pull automatically, correct?

Yes and no - it's generally not advised to use latest as it's not clear exactly how the infrastructure behaves. I think we could turn on continuous pre-pulling to reduce the chance of old images lying around, but you could end up with a node running the old image because it hasn't been told to pull a new one yet.

**Comment by khider** [comment]
I was advised to leave it at latest so we can update as needed. But I pushed a new container. Let me know if this solve the problem.

Re: logon. We will have people from multiple institutions, so not sure I can anticipate all the domains. I'm guessing this is why GitHub might work better unless we do something as generic as Google.

**Comment by sgibson91** [comment]
Ok, please let me know which GitHub org you would like to use and I will work on switching the auth method over.

**Comment by sgibson91** [comment]
> do I need to add everyone interested in using the hub to the organization?

Yes

**Comment by CommonClimate** [comment]
Hi! @khider will be on vacation next week so I'll try to fill in for her, with the caveat that I know precisely NOTHING about cloud things, but I am eager to learn!

**Comment by sgibson91** [comment]
I am an vacation next week too, so you will be in the capable hands of @yuvipanda!

**Comment by khider** [comment]
So I tried to login with my GitHub but was never asked to add LinkedEarth as an org. So it went to the 403:Forbidden screen. Any way I can add it directly into my org settings?

**Comment by sgibson91** [comment]
@khider I suspect this is because you tried to login when we were using CILogon. I'd try clearing your browser cookies and/or logging in from a private/incognito browser. The docs I linked to above also contain information on how to resolve the 403 error

* https://infrastructure.2i2c.org/en/latest/howto/configure/auth-management.html#follow-up-github-organization-administrators-must-grant-access

**Comment by yuvipanda** [comment]
I've requested a quota increase from Google Cloud as well, so you can expand to more users when necessary. I've asked for 256 total CPUs and 64 total nodes.

**Comment by khider** [comment]
Sorry all, was taking some time off. 

I was able to logon but I was never asked to allow LinkedEarth as an org. I'll try to add other people in the org that have the same level of power over authenticating.

On another note, I'm getting the following:

<img width="1281" alt="image" src="https://user-images.githubusercontent.com/11758571/178371940-16ba9b98-57b7-4da9-b5c7-9a34bb91ab06.png">

**Comment by khider** [comment]
Also, I'm able to login but not anyone from the GitHub organization. I was never asked to authenticate the app and it doesn't show on my GitHub apps page. Any advice on adding it manually?

**Comment by damianavila** [comment]
@khider, did you follow the advice from Sarah on this message: https://github.com/2i2c-org/infrastructure/issues/1418#issuecomment-1173598394?

@sgibson91, do you have any further thoughts besides the instructions you shared before?

**Comment by khider** [comment]
Yes. And that solved the issue for me. I was able to access it. But I can't add anyone else, they get a 403 permission error as well.

**Comment by khider** [comment]
Do I need to publicly list the members? I assume I can do that an as owner or is it locked to a user's preference?

**Comment by sgibson91** [comment]
@khider I changed the scope so it should work without people listing their membership publicly now. But each person would have to change it on their own profile, you wouldn't be able to do it for them.

**Comment by khider** [comment]
Thanks! Was able to get people to log in.

**Comment by damianavila** [comment]
Given the last message from @khider, I consider this one ready to be closed.
Thanks @sgibson91 for leading this effort.

If you have any other issues, @khider, feel free to reach out to the support email.

**Comment by khider** [comment]
Hi all, 

We have been getting ready to use the Hub more with users. Few questions I couldn't figure out:

1. I did select jupyterlab in the configurator but right now, the default view is notebook; not sure how to reset it. With that in mind, is it possible to install JupyterLab extensions?

2. We will start doing small scale tutorials. I tried to do one last week but was limited to 7 users. After that, the Hub wouldn't load. Is there a way for me to pre-warm nodes for these tutorials? They are 1hr in length but require a few people to login simultaneously. 

3. We're planning an event in June in which a group of 10-12 people will work together towards a research problem. It will be 5 days, will need to support 10-12 simultaneous users but not as tightly constrained as we have done before. Can we use the hub or should I consider getting an event hub for that week? I would love to be able to use this Hub since the idea is to help them get started on this research project throughout the week but they would keep working at their own pace later on.
